Hostname Guide V1 (#9762)
keycloak/keycloak: docs/guides/src/main/server/hostname.adoc

<#import "/templates/guide.adoc" as tmpl>
<#import "/templates/kc.adoc" as kc>
<#import "/templates/links.adoc" as links>

<@tmpl.guide
title="Configure the Hostname"

> summary="Learn how to use Keycloaks hostname configuration to expose Keycloaks endpoints the right way."

Reviewer: What about something like:

```suggestion
summary="Learn how to configure the frontend and backchannel endpoints exposed by Keycloak."
```

includedOptions="hostname-* proxy">

When running Keycloak in environments such as Kubernetes, OpenShift or even on-premise, you usually do not want to expose the internal URLs to the public facing internet. Instead, you want to use your public hostname. This guide will show you how to configure Keycloak to use the right hostname for different scenarios.

== Keycloak API Endpoint categories
Keycloak exposes three different API endpoint categories, each using a specific base URL. These categories are:

=== Frontend Endpoints
Frontend endpoints are used to externally access Keycloak. When no Hostname is set, the base URL used for the frontend is taken from the incoming request. This has some major disadvantages, e.g. in High availability setups spanning multiple Keycloak instances, where it should not depend on the instance the request lands what URL is used, but instead one URL should be used for all instances, so they are seen as one system from the outside.

The `hostname` property can be used to set the frontend base URL:

<@kc.start parameters="--hostname=<value>"/>

=== Backend Endpoints

> Backend endpoints are used for direct communication between Keycloak and applications, so normally without user interaction. Examples are the Token endpoint or the User info endpoint. Backend endpoints are also taking the base URL from the request by default. To override this behaviour, set the `hostname-strict-backchannel` configuration option:

Reviewer: Remove `so normally without user interaction`

<@kc.start parameters="--hostname=<value> --hostname-strict-backchannel=true"/>

> If set to `true`, all endpoints use the URL which was set as `hostname` value.

Reviewer: Should add a blurb about if all clients communicate through the public URL then this option should be enabled, but otherwise not having it enabled allows both internal and external applications

=== Administrative Endpoints
Keycloaks administration endpoints and its admin console should usually not be publicly accessible at all. You can force keycloak to use a specific, private URL for all `/admin/` endpoints:

<@kc.start parameters="--hostname=<publicHostname> --hostname-admin=<securedHostname>"/>

When `hostname-admin` is configured, all links and static resources used to e.g. render the admin console are served from the value you enter for `<securedHostname` instead of using `<publicHostname`.

> For more information about which paths to expose using a reverse proxy, please refer to the <@links.server id="proxy"/> Guide.

Reviewer: This is quite misleading as it can be interpreted as this somehow exposing the admin on a different URL. Should make it explicit that this isn't changing how they are exposed, just telling Keycloak where they are exposed.

== Overriding the hostname path

> When running behind a reverse proxy, you may expose keycloak using a different context root URL such as `myproxy.url/mykeycloak`. For this example, you can override the hostname path to use the path defined in your reverse proxy:

Reviewer: Keycloak, not keycloak

Reviewer: Different `context path` not `context root URL`?

<@kc.start parameters="--hostname=myurl --hostname-path=mykeycloak"/>

> When we assume  `hostname` and `hostname-path` are set up, when a HTTPS request arrives using either the hostname or the internal URL, all frontend endpoints served by keycloak would use `https://myurl/mykeycloak` as base URL instead, so that all traffic goes through your reverse proxy.

Reviewer: This sentences makes no sense to me ;)

> This only takes effect when a proxy mode different to `none` is used. Please see the <@links.server id="proxy"/> Guide for details.

Reviewer: `when a proxy is enabled`?

== Using the hostname in development mode

> When running Keycloak in development mode by using `start-dev`, the hostname setting is optional, so basically everything is taken from the incoming requests headers per default.

Reviewer: Avoid words like `so` and `basically`. Something like?
```suggestion
When running Keycloak in development mode by using `start-dev`, the hostname setting is optional. When the `hostname` is not specified the incoming request headers will be used.
```

=== Example: Hostname in development mode
.Keycloak configuration:
<@kc.startdev parameters=""/>

.Invoked command:
[source, bash]
----
curl GET "https://localhost:8080/realms/master/.well-known/openid-configuration" | jq .
----

.Result:
[source, bash]
----
# Frontend endpoints: request://request:request -> http://localhost:8080
# Backend endpoints: request://request:request -> http://localhost:8080
----

> invoking the curl GET request above when running in development mode will show you the current OpenID Discovery configuration. In this configuration, all base URLS will be taken from the incoming request, so `http://localhost:8080` will be used for all endpoints.

Reviewer: ```suggestion
Invoking the curl GET request above when running in development mode will show you the current OpenID Discovery configuration. In this configuration, all base URLS will be taken from the incoming request, so `http://localhost:8080` will be used for all endpoints.
```

== Example Scenarios
Here are a few more example scenarios and the corresponding commands for setting up a hostname.

=== Assumptions for all scenarios
* Keycloak is set up using HTTPS certificates and Port 8443.
* `intlUrl` refers to an internal IP/DNS for Keycloak

> * `myUrl` refers to an exposed external URL

Reviewer: exposed public URL?
